### ISSUE TITLE
Accesses to HID service in HoG sample dropped

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1958]
**_Reported by Jack Bradach:_**

Strange behavior running the HID-over-GATT (HoG) example on an Intel TinyTile.  Services enumerate in nRF Connect.  Reads to Device Info, Battery, and Generic Access characteristics work as expected.  Read from any characteristic under HID, and it never completes.

I turned on the BLE debug stream and watched with btmon.  It's acting like it doesn't see the read request to the HID Characteristics; the phone sends the read but it's like it's getting dropped before the debug stream would see it.  Once a read is dropped, I have to disconnect and reconnect before I can read any of the other (working) characteristics.  It doesn't seem to matter if I'm bonded or not.  I've attached a capture where I connect, read a characteristic from battery and devinfo services before trying to read one from HID.  The log is not truncated; the read simply never shows up.

I've been examining how the gatt/hog.c code is declaring the service compared to other working services (eg, gatt/bas.c, gatt/dis.c).  The most obvious difference is the length of the bt_gatt_attr being registered.  I cut it down to a single 'HIDS_INFO' characteristic to try to rule out the registration function goofing on larger lists; it didn't seem to make a difference (reads still hang).

Possibly unrelated, I'm remembering seeing similar behavior last week when trying to set up an HID device using the Arduino stack for this platform.  The google research I did on it led me to dismiss it as the Arduino implementation being rough (and switch to Zephyr), but that's the second implementation on this platform exhibiting the dropped reads.

Does anyone have a non-Intel Quark platform they could run the peripheral_hids example on?  If the HID service characteristics read correctly, that'd narrow the scope of this bug to the Intel targets.

(Imported from Jira ZEP-2117)

**Comment by nashif** [comment]
**_by Jack Bradach:_**

* [hid_timeout.btsnoop](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1958/hid_timeout.btsnoop) (size: 2841)

**Comment by nashif** [comment]
**_by Jack Bradach:_**

Similar behavior (with the reads failing), perhaps a bug on the same theme.

**Comment by nashif** [comment]
**_by Luiz Augusto von Dentz:_**

Are you guys sure this is not the tool failing for some reason, in the logs all seems fine:

< ACL Data TX: Handle 0 flags 0x00 dlen 24                           #10 [hci0] 0.731616
      ATT: Read By Group Type Response (0x11) len 19
        Attribute data length: 6
        Attribute group list: 3 entries
        Handle range: 0x0001-0x000b
        UUID: Human Interface Device (0x1812)
        Handle range: 0x000c-0x0010
        UUID: Generic Access Profile (0x1800)
        Handle range: 0x0011-0x0014
        UUID: Battery Service (0x180f)
...
> ACL Data RX: Handle 0 flags 0x02 dlen 11                           #40 [hci0] 1.371616
      ATT: Read By Type Request (0x08) len 6
        Handle range: 0x000c-0x0010
        Attribute type: Characteristic (0x2803)
< ACL Data TX: Handle 0 flags 0x00 dlen 20                           #41 [hci0] 1.371616
      ATT: Read By Type Response (0x09) len 15
        Attribute data length: 7
        Attribute data list: 2 entries
        Handle: 0x000d
        Value: 020e00002a
        Handle: 0x000f
        Value: 021000012a

Weird part is that there is no Find Info in the range of HID (000c-0010), also I curious what tool is this that does the discovery depth-first instead of breadth-first, it doesn't seem to do discovery for Secondary and does the Include discovery service by service.

**Comment by nashif** [comment]
**_by Jack Bradach:_**

No, nRF Connect is not ruled out, although I'll be surprised if it ends up being the root cause; they've got enough users that I'd expect a mention of a "nRF occasionally decides, 'nah, not feeling it' when told to read characteristics" would show up on Google. :-)

Also, I think I was able to reproduce this using the Bluez tools over a USB<->BT dongle (although I wasn't yet looking for a BLE bug so wasn't keeping great notes).  I'll recreate that data and attach it, as that'll give us a second implementation that's hitting this behavior (and reduce the chances that it's nRF Connect doing something incorrect).

**Comment by nashif** [comment]
**_by Jack Bradach:_**

_Carles Cufi_  pointed out that HID services is likely being claimed by the operating system, preventing nRF Connect from accessing them.  This appears to be the case, both in Linux and Android.

With that in mind, I was able to get the peripheral_hids example working on a Curie to talk to my Nexus 6p; I paired it using the normal Bluetooth widget rather than poking at it with nRF Connect and got a mouse pointer on the screen.  A quick 'notify' routine to update values, and I got it moving.

Closing as 'wontfix' as this isn't a Zephyr bug.  I have to do a little bit more research, but I'm planning to write some documentation on this behavior to include with the example.
